coriolis/osmorphing/osmount/base.py

```diff
@@ -298,6 +299,98 @@ def _get_volume_block_devices(self):
         LOG.info("Volume block devices: %s", volume_devs)
         return volume_devs
 
+    def _find_dev_with_contents(self, devices, all_files=None,
+                                one_of_files=None):
+        if all_files and one_of_files:
+            raise exception.CoriolisException(
+                "all_files and one_of_files are mutually exclusive")
+        dev_name = None
+        for dev_path in devices:
+            dirs = None
+            tmp_dir = self._exec_cmd('mktemp -d').decode().split('\n')[0]
+            try:
+                self._exec_cmd('sudo mount %s %s' % (dev_path, tmp_dir))
+                # NOTE: it's possible that the device was mounted successfully
+                # but an I/O error occurs later along the line:
+                dirs = self._exec_cmd('ls %s' % tmp_dir).decode().split('\n')
+            except Exception:
+                self._event_manager.progress_update(
+                    "Failed to mount and scan device '%s'" % dev_path)
+                LOG.warn(
+                    "Failed to mount and scan device '%s':\n%s",
+                    dev_path, utils.get_exception_details())
+                utils.ignore_exceptions(self._exec_cmd)(
+                    "sudo umount %s" % tmp_dir
+                )
+                utils.ignore_exceptions(self._exec_cmd)(
+                    "sudo rmdir %s" % tmp_dir
+                )
+                continue
+
+            LOG.debug("Contents of device %s:\n%s", dev_path, dirs)
+
+            if all_files and dirs:
+                common = [i if i in dirs else None for i in all_files]
+                if not all(common):
+                    self._exec_cmd('sudo umount %s' % tmp_dir)
+                    continue
+
+                dev_name = dev_path
+                self._exec_cmd('sudo umount %s' % tmp_dir)
+            elif one_of_files and dirs:
+                common = [i for i in one_of_files if i in dirs]
+                if len(common) > 0:
+                    dev_name = dev_path
+                    self._exec_cmd('sudo umount %s' % tmp_dir)
+                    break
+            else:
+                self._exec_cmd('sudo umount %s' % tmp_dir)
+                continue
+
+        if dev_path:
```

Comment:
1) I appreciate the idea of removing devices we've successfully identified so as not to pointlessly rescan them but IMO mutating an argument is not ideal and we should avoid it if possible (IMO the like 3s extra wait to rescan the volume is not that big of a deal in the grand scheme of things)
2) even if you decide against it, please declare a separate variable for the `dev_path_to_remove` and pop that away as currently you'll be popping the device last iterated over if and `all_files` arg is provided (there's no break in the first `if`)

Reply:
The device is removed if, and only if one is found, prior to returning that particular value at the end of this function. Not all devices get removed.

The variable ```all_files``` is supposed to not break. It represents an all or nothing condition. Either all expected files are present, or the device is not relevant and we go on to check the next one in the list.

This ```if dev_path``` is outside of that loop, and ```dev_path``` only gets set to something other than ```None``` if a valid device is found. If none are found, nothing gets removed.

When one is found, it is returned as a value, which can later be used by the caller as a parameter in whichever manner they wish.

I can remove this, and add it to ```_find_and_mount_root()```. That function actually mounts the device, and there is no reason to keep that device in the list and risk mounting it twice, leading to potential cases where we forget about the second mountpoint and fail to ```umount``` it.

Reply:
Well either way this code will remove the last `dev_path` the `for` iterated over (which isn't necessarily the one we actually found), not to mention that we'll get a `NameError` if `devices` is an empty list (which we could just check for at the top if deemed necessary)
I realise now that you were probably meaning to check and remove the `dev_name` variable you return below, not `dev_path`.

Reply:
Moved to ```_find_and_mount_root()```